pkg/services/annotations/annotationsimpl/loki/alert_state_history_store.go

```diff
@@ -0,0 +1,399 @@
+				continue
+			}
+
+			if !shouldReplay(entry, transition, resources) {
+				continue
+			}
+
+			rule, ok := rules[entry.RuleUID]
+			if !ok {
+				ruleQ := ruleQuery{OrgID: orgID, UID: entry.RuleUID}
+				rule, err = r.getRule(ctx, ruleQ)
+				if err != nil {
+					r.log.Debug(fmt.Sprintf("could not find rule with UID `%s`", entry.RuleUID), "entry", entry, "err", err)
+					continue
+				}
+
+				rules[entry.RuleUID] = rule
+			}
+
+			item, err := buildAnnotationItem(&entry, resources.Dashboards[entry.DashboardUID], rule, transition.State)
+			if err != nil {
+				r.log.Debug("could not build annotation item", "entry", entry, "err", err)
+				continue
+			}
+
+			items = append(items, item)
+		}
+	}
+	return items
+}
+
+type ruleQuery struct {
+	OrgID int64
+	UID   string
+	ID    int64
+}
+
+func (r *AlertStateHistoryStore) getRule(ctx context.Context, query ruleQuery) (*ngmodels.AlertRule, error) {
+	rule := &ngmodels.AlertRule{OrgID: query.OrgID}
+	if query.UID != "" {
+		rule.UID = query.UID
+	}
+	if query.ID != 0 {
+		rule.ID = query.ID
+	}
+
+	err := r.db.WithDbSession(ctx, func(sess *db.Session) error {
+		exists, err := sess.Get(rule)
+		if err != nil {
+			return err
+		}
+		if !exists {
+			return ErrMissingRule
+		}
+		return nil
+	})
+
+	return rule, err
+}
+
+// This store is read-only, so no other methods are implemented
+
+func (r *AlertStateHistoryStore) Add(cxt context.Context, item *annotations.Item) error {
+	return ErrLokiStoreUnsupported
+}
+
+func (r *AlertStateHistoryStore) AddMany(ctx context.Context, items []annotations.Item) error {
+	return ErrLokiStoreUnsupported
+}
+
+func (r *AlertStateHistoryStore) Update(ctx context.Context, item *annotations.Item) error {
+	return ErrLokiStoreUnsupported
+}
+
+func (r *AlertStateHistoryStore) Delete(ctx context.Context, params *annotations.DeleteParams) error {
+	return ErrLokiStoreUnsupported
+}
+
+func (r *AlertStateHistoryStore) GetTags(ctx context.Context, query *annotations.TagsQuery) (annotations.FindTagsResult, error) {
+	return annotations.FindTagsResult{}, ErrLokiStoreUnsupported
+}
+
+func (r *AlertStateHistoryStore) CleanAnnotations(ctx context.Context, cfg setting.AnnotationCleanupSettings, annotationType string) (int64, error) {
+	return 0, ErrLokiStoreUnsupported
+}
+
+func (r *AlertStateHistoryStore) CleanOrphanedAnnotationTags(ctx context.Context) (int64, error) {
+	return 0, ErrLokiStoreUnsupported
+}
+
+// util
+
+func shouldReplay(entry historian.LokiEntry, transition *state.StateTransition, resources accesscontrol.AccessResources) bool {
+	if !historian.ShouldRecord(*transition) {
+		return false
+	}
+
+	if _, ok := resources.ScopeTypes[annotations.Organization.String()]; ok {
+		if entry.DashboardUID != "" {
+			return false
+		}
+	}
+
+	if _, ok := resources.ScopeTypes[annotations.Dashboard.String()]; ok {
+		_, canAccess := resources.Dashboards[entry.DashboardUID]
+		if !canAccess {
+			return false
+		}
+	}
```

Comment:
Nit: Could we separate the RBAC filter from the business logic filter of `ShouldRecord`? These filters effectively come from different layers of the application, and are subject to change for different reasons.

Even just a separate function would make this more clear, it would cleanly mark which code is implementing the existing annotation rbac.

Reply:
Good point, these two pieces really aren't related enough to warrant being grouped like this.